File: WebApp/src/common/wait.ts
import sleep from "./sleep"

interface Options {
  minMs: number
}

/// Waits for a promise or at least the given milliseconds.
export default async function wait<T>(promise: Promise<T>, options: Options): Promise<T> {
  const startTime = Date.now()

  async function sleepIfNeeded() {
    const elapsedTime = Date.now() - startTime
    const remainingTime = options.minMs - elapsedTime
    await sleep(remainingTime)
  }

  try {
    const result = await promise
    sleepIfNeeded()
    return result

  } catch(error) {
    sleepIfNeeded()
    throw error
  }
}

File: iOS-Quiz/Sources/Utils/Wait.swift
import Foundation

func wait<R>(atLeast: Duration, _ closure: () async throws -> R) async throws -> R {
    let clock = ContinuousClock()
    let startTime = clock.now

    let result: Result<R, Error> = await {
        do {
            return .success(try await closure())
            
        } catch {
            return .failure(error)
        }
    }()
    
    let elapsedTime = clock.now - startTime
    let remainingTime = atLeast - elapsedTime
    
    if remainingTime > .zero {
        try? await Task.sleep(for: remainingTime)
    }
    
    switch result {
    case .success(let value):
        return value
        
    case .failure(let error):
        throw error
    }
}
